### PR TITLE
CDI 3.0 FATs

### LINKED ISSUE #15517: Run CDI FAT buckets against cdi-3.0
Currently only some FAT buckets are properly transformed to run against cdi-3.0. We need to transform them all. We also need to ensure that all tests in each bucket are repeated against cdi-3.0. As with other components, we need to establish the correct balance of running tests against every version and also LITE vs FULL.

Every test should be run against the latest version.
In most cases, running against cdi-3.0 is sufficiently similar to cdi-2.0 such that a repeat is not required.
Some tests should be run against cdi-1.2 (but probably in FULL mode only).
Of all of those test runs, golden path tests should be in LITE mode.
Edge cases and failure tests should be in FULL mode.

These are the buckets that have been transformed. Even in the transformed ones, not all of the tests are properly repeated.
- [x] com.ibm.ws.cdi.annotations_fat
- [x] com.ibm.ws.cdi.api_fat
- [x] com.ibm.ws.cdi.beansxml_fat
- [x] com.ibm.ws.cdi.beansxml.implicit_fat
- [x] com.ibm.ws.cdi.ejb_fat
- [x] com.ibm.ws.cdi.extension_fat
- [x] com.ibm.ws.cdi.jee_fat
- [x] com.ibm.ws.cdi.jndi_fat
- [x] com.ibm.ws.cdi.lifecycle_fat
- [x] com.ibm.ws.cdi.noncontextual_fat
- [x] com.ibm.ws.cdi.third.party_fat
- [x] com.ibm.ws.cdi.visibility_fat

### PR DESCRIPTION
**Pull request by tevans78**
- Update CDI FAT buckets to repeat against JEE9
- Use standard FAT framework wherever possible
- Simplify FAT project layout
- Remove redundant code
- Ensure that all tests are properly enabled

fixes #15517

### CODE REVIEW

**Comment by tevans78** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

File: dev/com.ibm.ws.cdi.beansxml.implicit_fat/fat/src/com/ibm/ws/cdi/beansxml/implicit/tests/ImplicitWarLibJarsTest.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        JavaArchive implicitBeanExplicitArchive = ShrinkWrap.create(JavaArchive.class, "implicitBeanExplicitArchive.jar")
+                                                            .addClass(com.ibm.ws.cdi.beansxml.implicit.apps.beans.InExplicitBeanArchive.class)
+                                                            .add(new FileAsset(new File("test-applications/implicitBeanExplicitArchive.jar/resources/META-INF/beans.xml")),
+                                                                 "/META-INF/beans.xml");
+
+        JavaArchive utilLib = ShrinkWrap.create(JavaArchive.class, "utilLib.jar")
+                                        .addClass(com.ibm.ws.cdi.beansxml.implicit.utils.ChainableListImpl.class)
+                                        .addClass(com.ibm.ws.cdi.beansxml.implicit.utils.Intercepted.class)
+                                        .addClass(com.ibm.ws.cdi.beansxml.implicit.utils.ChainableList.class)
+                                        //.addClass(com.ibm.ws.cdi.beansxml.implicit.utils.Utils.class)
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.beansxml.implicit_fat/fat/src/com/ibm/ws/cdi/beansxml/implicit/tests/ImplicitWarLibJarsTest.java:82]: Random commented out line

**Comment by Azquelt** [comment]
Having just reviewed beansxml.implicit_fat, it would probably have been preferable to put classes into packages by test, rather than grouping them by beans, ejbs, servlets, other.

Still a vast improvement over the old code, however.

File: dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/fat/tests/EnablingBeansXmlValidationTest.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ public static void setUp() throws Exception {
      * Test to ensure that an old beans.xml which was parsed by OWB is rejected as long as validation is enabled.
      */
     @Test
-    @SkipForRepeat(SkipForRepeat.EE8_FEATURES)
+    @SkipForRepeat({SkipForRepeat.EE8_FEATURES, SkipForRepeat.EE9_FEATURES})
     @AllowedFFDC({ "org.jboss.weld.exceptions.IllegalStateException", "com.ibm.ws.container.service.state.StateChangeException" }) //We are expecting these errors, but we're using Allowed because the schema download occasionally fails.
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/fat/tests/EnablingBeansXmlValidationTest.java:60]: This `@SkipForRepeat` should have a comment explaining why it's not tested on EE8/9.

I.e. is it just that it's not worth repeating, or is it not valid on 8/9?

File: dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java [hunk 3]
```diff
@@ -48,7 +48,7 @@
 public class FATSuite {
 
     @ClassRule
-    public static RepeatTests r = RepeatTests.withoutModification().andWith(FeatureReplacementAction.EE8_FEATURES());
+    public static RepeatTests r = RepeatTests.withoutModification().andWith(FeatureReplacementAction.EE8_FEATURES()).andWith(FeatureReplacementAction.EE9_FEATURES());
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java:51]: Is it correct to run against all versions in LITE mode?

File: dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/aroundconstruct/AroundConstructTestServlet.java [hunk 4]
```diff
@@ -100,6 +136,11 @@ public void testInterceptorOrder() {
 
     }
 
+    /**
+     * Interceptors should only be called once for each constructor
+     */
+    @Test
+    @Ignore //TODO need to find out why this test does not pass
     public void testInterceptorNotCalledTwice() {
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/aroundconstruct/AroundConstructTestServlet.java:144]: Is there an issue for this?

Comment (tevans78) [reply dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/aroundconstruct/AroundConstructTestServlet.java:144]: https://github.com/OpenLiberty/open-liberty/issues/15711

File: dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java (hunk 3 above)
Comment (tevans78) [reply dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java:51]: No ... plan to do a follow up PR to continue tidying/refactoring

File: dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/fat/tests/EnablingBeansXmlValidationTest.java (hunk 2 above)
Comment (tevans78) [reply dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/fat/tests/EnablingBeansXmlValidationTest.java:60]: I'll let @benjamin-confino answer that one, I don't know

File: dev/com.ibm.ws.cdi.beansxml.implicit_fat/fat/src/com/ibm/ws/cdi/beansxml/implicit/tests/ImplicitWarLibJarsTest.java (hunk 1 above)
Comment (tevans78) [reply dev/com.ibm.ws.cdi.beansxml.implicit_fat/fat/src/com/ibm/ws/cdi/beansxml/implicit/tests/ImplicitWarLibJarsTest.java:82]: Line can be deleted at some point.

File: dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/postConstructError/ErrorMessageServlet.java [hunk 5]
```diff
@@ -28,7 +30,11 @@
 
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+System.out.println("GREP");
         errEjb.doSomething();
+        PrintWriter pw = response.getWriter();
+        pw.write("Hello World");
+System.out.println("GREP2");
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/postConstructError/ErrorMessageServlet.java:37]: Left over debugging sysouts?

File: dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/EjbTimerTest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+        //first couple of times is synchronous (no timer or task)
+        verifyResponse(wb, server, "/ejbTimer/NoTimer", "session = 0 request = 0");
+        verifyResponse(wb, server, "/ejbTimer/NoTimer", "session = 1 request = 0");
+
+        //the next couple start a timer which will increment asynchronously after 1 second
+        //only one timer can be active at a time so subsequent calls will block until the previous timers have finished
+        verifyResponse(wb, server, "/ejbTimer/Timer", "session = 2 request = 0");
+        verifyResponse(wb, server, "/ejbTimer/Timer", "session = 3 request = 0");
+        verifyResponse(wb, server, "/ejbTimer/NoTimer", "session = 4 request = 0");
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/EjbTimerTest.java:105]: Good news: it was reasonably easy to find the classes which implement this test.

Bad news: the call to `/ejbTimer/Timer` appears to increment the session scoped counter after setting up a timer, rather than when that timer executes, contrary to the javadoc on this test method. (See `EjbSessionBean#incCountersViaTimer`)

File: dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.war/src/com/ibm/ws/cdi12/test/jndi/BeanManagerP.java [hunk 7]
```diff
@@ -8,28 +8,30 @@
...
-        resp.getOutputStream().println(TestUtils.join(ejb.getResult()));
-    }
+@ApplicationScoped
+public class BeanManagerP  {
+
+    @Produces
+    @Resource(lookup = "java:comp/BeanManager")
+    private String manager;
 
 }
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.war/src/com/ibm/ws/cdi12/test/jndi/BeanManagerP.java:37]: This appears to be a new class (even though github is showing it as having moved from another project). What is it for? I don't see a corresponding change in this project where a test would make use of it.

File: dev/com.ibm.ws.cdi.third.party_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2018, 2020 IBM Corporation and others.
+ * Copyright (c) 2018, 2021 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.third.party_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java:5]: The copyright header has been changed but there are no other changes to the class. Should the `RepeatTests` rule have been updated?

File: dev/com.ibm.ws.cdi.visibility_fat/publish/servers/cdi12BasicServer/server.xml [hunk 9]
```diff
@@ -5,6 +5,7 @@
     <featureManager>
         <feature>osgiconsole-1.0</feature>
         <feature>cdi-1.2</feature>
+        <feature>jaxrs-2.0</feature>
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.visibility_fat/publish/servers/cdi12BasicServer/server.xml:8]: Why is this needed?

File: dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java [hunk 10]
```diff
@@ -53,4 +71,14 @@ public static void generateHelpFile() {
...
 
+    @AfterClass
+    public static void transformUserFeatures() {
+        //If we've finished EE8 then transform javax to Jakarta in preperation for EE9
+        if (RepeatTestFilter.isRepeatActionActive("EE8_FEATURES")) {
+            for (String path : PATHS_TO_BUNDLES) {
+                JakartaEE9Action.transformApp(Paths.get(path));
+            }
+        }
+    }
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java:82]: I hate this.

Ideally, ensure that two versions of each bundle get built, one for javax and one for jakarta so that both can be deployed at once and the test order can be changed without breaking the tests.

At the very least please put it in a `@BeforeClass` when we're starting the EE9 features, rather than relying on EE9 always following EE8.

**Review by Azquelt — COMMENTED** [review]
I've reviewed all files, please see the comments above.

File: dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/EjbTimerTest.java (hunk 6 above)
Comment (tevans78) [reply dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/EjbTimerTest.java:105]: I don't think that aspect of the test has changed ... so an issue should be opened to fix it but shouldn't stop this PR.

Comment (tevans78) [reply dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/EjbTimerTest.java:105]: https://github.com/OpenLiberty/open-liberty/issues/15789

File: dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.war/src/com/ibm/ws/cdi12/test/jndi/BeanManagerP.java (hunk 7 above)
Comment (tevans78) [reply dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.war/src/com/ibm/ws/cdi12/test/jndi/BeanManagerP.java:37]: Agree, this looks misplaced? @benjamin-confino

File: dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/fat/tests/EnablingBeansXmlValidationTest.java (hunk 2 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.beansxml_fat/fat/src/com/ibm/ws/cdi12/fat/tests/EnablingBeansXmlValidationTest.java:60]: Weld changed some of it's output so we have to versions of this test. One for CDI 1.2 and one for CDI 2+

The other method has the explaining comment: `//In CDI 2.0 with weld 3.0.4 or later WELD-001210 is a warning not an error. `

File: dev/com.ibm.ws.cdi.visibility_fat/publish/servers/cdi12BasicServer/server.xml (hunk 9 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.visibility_fat/publish/servers/cdi12BasicServer/server.xml:8]: I don't remember for certain, but I suspect it has something to do with this comment from the deleted WarLibsAccessWarBeansTest `//This test should be merged into BasicVisibilityTests when jaxrs-3.0 (EE9) is available`

jaxrs-3.0 is available, the tests were merged, and I suspect that during that process I hit an error that was resolved by explicitly listing jaxrs-2.0 in the features.

File: dev/com.ibm.ws.cdi.third.party_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java (hunk 8 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.third.party_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java:5]: Most likely.

File: dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.war/src/com/ibm/ws/cdi12/test/jndi/BeanManagerP.java (hunk 7 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.war/src/com/ibm/ws/cdi12/test/jndi/BeanManagerP.java:37]: I agree. It can be deleted.

File: dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java (hunk 10 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java]: I really don't like the idea of having two identical bundles that must always be kept in sync, so I moved it to the BeforeClass.

File: dev/com.ibm.ws.cdi.third.party_fat/fat/src/com/ibm/ws/cdi12/fat/tests/HibernateSearchTest.java [hunk 11]
```diff
@@ -46,7 +47,25 @@
 import componenttest.rules.repeater.EERepeatTests;
 import componenttest.topology.utils.FATServletClient;
 
+/*
+
+The following error occurs on EE9. When this is fixed remove the skip for repeat. 
+
+Caused by: java.lang.NoClassDefFoundError: javax.persistence.spi.PersistenceProvider
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.third.party_fat/fat/src/com/ibm/ws/cdi12/fat/tests/HibernateSearchTest.java:54]: Is there an issue raised for this? (Had a quick look but couldn't see one)

File: dev/com.ibm.ws.cdi.visibility_fat/publish/servers/cdi12BasicServer/server.xml (hunk 9 above)
Comment (Azquelt) [reply dev/com.ibm.ws.cdi.visibility_fat/publish/servers/cdi12BasicServer/server.xml:8]: Hmm, I can't see anything in the test that requires jaxrs at all.

I tried replacing jaxrs-2.0 with servlet-3.1 in the original test and it still passes so I don't think it was ever needed by the test.

In addition, it looks like the `cdi12WarLibsAccessWarServer` server can be removed.

File: dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java (hunk 10 above)
Comment (Azquelt) [reply dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi12/suite/FATSuite.java]: Ah, I wasn't intending that we'd have two copies of the code, just that we'd compile it once as the EE8 bundle, then transform it into a separate EE9 bundle.

I'm not entirely sure of the build magic that would be required to make that happen, but it seems like it would be similar to the transformation that we do for product bundles.

**Review by Azquelt — COMMENTED** [review]
Two outstanding questions, everything else looks fine.

**Comment by benjamin-confino** [comment]
#build

**Review by Azquelt — APPROVED** [review]

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#run-libby-bot
